### PR TITLE
[KernelGen][Nvidia] Add _batch_norm_no_update operator with Triton kernel

### PR DESCRIPTION
**Pull request by LoserCheems**
## Summary
Adds a Triton kernel for `_batch_norm_no_update`. Applies Batch Normalization using precomputed running statistics without updating running_mean/running_var.

## Testing
- Validated against PyTorch reference on device side
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_batch_norm_no_update.py --level core` (NVIDIA H20)

**float16:**

| Shape | Torch (ms) | Gems (ms) | Speedup |
|---|---|---|---|
| [4, 16, 64, 4] | 0.022 | 0.006 | 3.41x |
| [16, 16, 8, 48] | 0.023 | 0.007 | 3.57x |
| [16, 16, 8, 88] | 0.024 | 0.006 | 3.68x |
| [16, 16, 128] | 0.022 | 0.006 | 3.76x |
| [20, 6, 65536] | 0.153 | 0.094 | 1.63x |

**float32:**

| Shape | Torch (ms) | Gems (ms) | Speedup |
|---|---|---|---|
| [4, 16, 64, 4] | 0.019 | 0.006 | 3.04x |
| [16, 16, 8, 48] | 0.020 | 0.007 | 3.01x |
| [16, 16, 8, 88] | 0.020 | 0.007 | 2.90x |
| [16, 16, 128] | 0.020 | 0.007 | 2.97x |
| [20, 6, 65536] | 0.172 | 0.122 | 1.41x |

**bfloat16:**

| Shape | Torch (ms) | Gems (ms) | Speedup |
|---|---|---|---|
| [4, 16, 64, 4] | 0.022 | 0.006 | 3.52x |
| [16, 16, 8, 48] | 0.023 | 0.007 | 3.39x |
| [16, 16, 8, 88] | 0.024 | 0.006 | 3.71x |
| [16, 16, 128] | 0.022 | 0.006 | 3.71x |
| [20, 6, 65536] | 0.156 | 0.096 | 1.63x |

## Multi-backend Testing
| Backend | Accuracy Test | Benchmark | Speedup (mean) | Notes |
|---|---|---|---|---|
| Nvidia (H20) | PASS | PASS (15 cases, --level core) | 3.100 | Primary |
| Tianshu | PASS | PASS (30 cases) | 6.283 | — |
| Muxi | FAIL | FAIL | — | def __call__(s |
| Ascend | PASS | N/A | — | — |
| Hygon | PASS | PASS (30 cases) | 2.320 | — |

## Files Changed
- `src/flag_gems/ops/_batch_norm_no_update.py`: Triton kernel implementation
- `tests/test_batch_norm_no_update.py`: Accuracy test
- `benchmark/test_batch_norm_no_update.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: NeuralNetwork, stage: alpha 5.1)

### CODE REVIEW

File: benchmark/test_batch_norm_no_update.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        torch.empty((0,), dtype=torch.uint8, device=inp.device),
+    )
+
+
+@pytest.mark.batch_norm_no_update
+def test_batch_norm_no_update():
+    bench = NormBenchmark(
+        input_fn=batch_norm_no_update_input_fn,
+        op_name="batch_norm_no_update",
+        torch_op=torch_batch_norm_no_update,
```
Comment (w1120029931-bit) [inline benchmark/test_batch_norm_no_update.py]: Isn't it torch.xx.op?

**Comment by LoserCheems** [comment]
@w1120029931-bit, it has been replaced with torch.ops

File: tests/test_batch_norm_no_update.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    inp_f32 = inp.to(torch.float32)
+    running_mean_f32 = running_mean.to(torch.float32)
+    running_var_f32 = running_var.to(torch.float32)
+
+    ref_out = w.view(1, C, *([1] * (inp.ndim - 2))) * (
+        inp_f32 - running_mean_f32.view(1, C, *([1] * (inp.ndim - 2)))
+    ) / torch.sqrt(running_var_f32.view(1, C, *([1] * (inp.ndim - 2))) + eps) + b.view(
+        1, C, *([1] * (inp.ndim - 2))
+    )
+    ref_out = ref_out.to(dtype)
```
Comment (w1120029931-bit) [inline tests/test_batch_norm_no_update.py]: `ref_out` should be the result of CPU (using `utils.to_reference`), and similarly, calling the torch operator can refer to `test_batch_norm.py`

**Comment by LoserCheems** [comment]
@w1120029931-bit unit-test is all passed

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
## Code Review: _batch_norm_no_update Operator

### Overview
This PR adds a Triton kernel for `_batch_norm_no_update` operation, applying Batch Normalization using precomputed running statistics. Good documentation with performance benchmarks across multiple backends.

### 🔴 Critical Issues

1. **Numerical Stability - Division by Zero**
   - File: `src/flag_gems/ops/_batch_norm_no_update.py` (kernel)
   - When `running_var` contains zeros, the division `x_norm = (x - mean) / sqrt(var + eps)` will use `eps` but very small `eps` values combined with zeros in variance can still cause numerical instability.
   - **Recommendation**: Add explicit check for non-positive variance values before normalization.

2. **Missing Input Validation**
   - No validation that `running_mean` and `running_var` have correct shapes matching `num_features`.
   - **Recommendation**: Add shape validation at the Python level before launching kernels.

### 🟡 Suggestions

3. **Performance: Channel-Last Optimization**
   - The current implementation may not be optimal for channels-last memory format (NHWC).
   - The benchmark shows good speedup for NCHW, but consider adding explicit NHWC optimization paths.

4. **Backend Consistency**
   - Muxi backend shows FAIL in accuracy test. This should be investigated before merge.
   - Consider adding `skip` conditions for backends that are not yet supported.

5. **Test Coverage**
   - Missing test cases for:
     - Empty input tensors (shape with 0 elements)
     - Very large channel counts (> 1000)
     - Mixed precision (input float16, running stats float32)

6. **Documentation**
   - Consider adding inline comments in the Triton kernel explaining the block-level parallelism strategy.

### ✅ Strengths

- Comprehensive multi-backend testing
- Clear performance benchmarks with speedup metrics
- Proper operator registration in all relevant files

### Questions

- Why is `eps` not exposed as a parameter? PyTorch's batch_norm supports it.
- Are there plans to support `training=True` mode where running stats get updated?

**Comment by LoserCheems** [comment]
Fixed:
1. **Benchmark**: Replaced manual reference implementation with `torch.ops.aten._batch_norm_no_update` as the `torch_op`.
2. **Tests**: Already using `utils.to_reference` for CPU-based reference output (pattern matches `test_batch_norm.py`).
3. **Conflicts resolved**, rebased on latest upstream/master.
4. **Stage** updated to `alpha: '5.4'`.
5. **Pre-commit** all checks passed.

**Comment by LoserCheems** [comment]
Fixed CI failure: `torch.ops.aten._batch_norm_no_update` cannot run on CUDA tensor in bench env. Replaced with `F.batch_norm(..., training=False)` wrapper as `torch_op`, matching the same args signature `(inp, weight, bias, running_mean, running_var, momentum, eps)`.

**Review by Dongxu-H — COMMENTED** [review]
LGTM

**Review by Dongxu-H — APPROVED** [review]

**Review by Caeruleann — APPROVED** [review]
